commit/merkleroot/processor.go

```diff
@@ -0,0 +1,83 @@
+package merkleroot
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	"github.com/smartcontractkit/chainlink-ccip/internal/reader"
+	"github.com/smartcontractkit/chainlink-ccip/pluginconfig"
+	"github.com/smartcontractkit/chainlink-ccip/shared"
+)
+
+// Processor is the processor responsible for
+// reading next messages and building merkle roots for them,
+// It's setup to use RMN to query which messages to include in the merkle root and ensures
+// the newly built merkle roots are the same as RMN roots.
+type Processor struct {
+	nodeID       commontypes.OracleID
+	cfg          pluginconfig.CommitPluginConfig
+	lggr         logger.Logger
+	observer     Observer
+	readerSyncer *plugincommon.BackgroundReaderSyncer
+	ccipReader   reader.CCIP
+	reportingCfg ocr3types.ReportingPluginConfig
+	chainSupport shared.ChainSupport
+}
+
+// NewProcessor creates a new Processor
+func NewProcessor(
+	nodeID commontypes.OracleID,
+	lggr logger.Logger,
+	cfg pluginconfig.CommitPluginConfig,
+	homeChain reader.HomeChain,
+	ccipReader reader.CCIP,
+	readerSyncer *plugincommon.BackgroundReaderSyncer,
+	msgHasher cciptypes.MessageHasher,
+	reportingCfg ocr3types.ReportingPluginConfig,
+	chainSupport shared.ChainSupport,
+) *Processor {
+	observer := ObserverImpl{
+		lggr,
+		homeChain,
+		nodeID,
+		chainSupport,
+		ccipReader,
+		msgHasher,
+	}
+	return &Processor{
+		nodeID:       nodeID,
+		cfg:          cfg,
+		lggr:         lggr,
+		ccipReader:   ccipReader,
+		readerSyncer: readerSyncer,
+		observer:     observer,
+		reportingCfg: reportingCfg,
+		chainSupport: chainSupport,
+	}
+}
+
+func (w *Processor) Close() error {
+	timeout := 10 * time.Second
+	ctx, cf := context.WithTimeout(context.Background(), timeout)
+	defer cf()
+
+	if err := w.readerSyncer.Close(); err != nil {
+		w.lggr.Errorw("error closing reader syncer", "err", err)
+	}
+
+	if err := w.ccipReader.Close(ctx); err != nil {
+		return fmt.Errorf("close ccip reader: %w", err)
+	}
+
+	return nil
+}
```

Comment:
These resources are already managed in the primary plugin. Also, is close ever called?

Reply:
good catch. I think we can remove it from here. Do I need to initialize them as references in processor in this case?

Reply:
Hmm, we should consider that a little further. They're now being used concurrently as well. Are they all thread safe?

Reply:
We removed the concurrency for simplicity now.

Reply:
We could add `Close` to the interface and call it from the root plugin